fedora_messaging/twisted/protocol.py

```diff
@@ -265,16 +254,215 @@ def publish(self, message, exchange):
+            self._channel = yield self._allocate_channel()
+            yield self.publish(message, exchange)
+        except pika.exceptions.ConnectionClosed as e:
+            raise ConnectionException(reason=e)
+
+    @defer.inlineCallbacks
+    def consume(self, callback, queue):
+        """
+        Register a message consumer that executes the provided callback when
+        messages are received.
+
+        The queue must exist prior to calling this method.  If a consumer
+        already exists for the given queue, the callback is simply updated and
+        any new messages for that consumer use the new callback.
+
+        If :meth:`resumeProducing` has not been called when this method is called,
+        it will be called for you.
+
+        Returns:
+            Consumer: A namedtuple that identifies this consumer.
+        """
+        if queue in self._consumers and self._consumers[queue].channel.is_open:
+            consumer = Consumer(
+                tag=self._consumers[queue].tag,
+                queue=queue,
+                callback=callback,
+                channel=self._consumers[queue].channel,
+            )
+            self._consumers[queue] = consumer
+            defer.returnValue(consumer)
+
+        consumer = Consumer(
+            tag=str(uuid.uuid4()), queue=queue, callback=callback, channel=self._channel
         )
 
+        if not self._running:
+            self._consumers[queue] = consumer
+            yield self.resumeProducing()
+            defer.returnValue(consumer)
+
+        self._consumers[queue] = consumer
+        queue_object, _ = yield consumer.channel.basic_consume(
+            queue=consumer.queue, consumer_tag=consumer.tag
+        )
+        deferred = self._read(queue_object, consumer)
+        deferred.addErrback(
+            lambda f: _log.failure, "_read failed on consumer {c}", c=consumer
+        )
+        _log.info("Successfully registered AMQP consumer {c}", c=consumer)
+        defer.returnValue(consumer)
+
+    @defer.inlineCallbacks
+    def cancel(self, queue):
+        """
+        Cancel the consumer for a queue.
+
+        Args:
+            queue (str): The name of the queue the consumer is subscribed to.
+
+        Returns:
+            defer.Deferred or None: Either a Deferred that fires when the consumer
+                is canceled, or None if the consumer was already canceled. Wrap
+                the call in :func:`defer.maybeDeferred` to always receive a Deferred.
+        """
+        try:
+            consumer = self._consumers[queue]
+            yield consumer.channel.basic_cancel(consumer_tag=consumer.tag)
+        except pika.exceptions.AMQPChannelError:
+            # Consumers are tied to channels, so if this channel is dead the
+            # consumer should already be canceled (and we can't get to it anyway)
+            pass
+        except KeyError:
+            return
+
+        try:
+            yield consumer.channel.close()
```

Comment:
Until every consumer is in their own channel, this will close the main channel.

Reply:
@jeremycline what do you think of that comment?

Reply:
Ah, each consumer now has its own channel, allocated in the ``consume()`` call.

Reply:
Do they? I'm not seing this, could you point me to the right line?

Reply:
Ah, my bad, the Github UI was somehow showing an old version.